### PR TITLE
virtual crystal for ml_norm 3d

### PR DESCRIPTION
**Pull request by gefeichen**
new_max_delta in ml_norm maybe not correct. 
when reviewing, ignore whitespace and empty lines should be chosen.

### CODE REVIEW

**Review by KrisThielemans — CHANGES_REQUESTED** [review]
thanks. some detailed comments. 2 general ones.

- I'd rather not introduce a lot of white-space differences in this PR. I'd rather do that by fixing #98 (see #724). Would it be possible to revert your white-space differences, or we wait till #724 is merged (could take a bit)
- I'd rather replace `make_fan_data` code `with the one from `make_fan_data_remove_gaps` etc and remove the explicit `*remove_gaps` functions. That'd mean we can only do this if the `*remove_gaps` functions give exactly the same result if there are no virtual crystals. We have no tests for anything like this unfortunately so I guess you have to do it by hand or even create a small test first! (create text files with norm factors, create some projdata (e.g. just 1), find norm factors, check)

**Comment by KrisThielemans** [comment]
@gefeichen any chance you can have another look at this?

**Comment by gefeichen** [comment]
> look

sorry for the later reply. I will finish it by December

**Comment by gefeichen** [comment]
> @gefeichen any chance you can have another look at this?



> @gefeichen any chance you can have another look at this?

sorry for the later reply. I will finish it by December

File: src/buildblock/ML_norm.cxx
Comment (gefeichen) [reply src/buildblock/ML_norm.cxx:1104]: The position of virtual crystal should be considered and I have another question: if the number of virtual crystal is 0 or 1, it's not necessary to set this offset.

 12(transaxial crystal)+3(virtual crystal) = 15

if index is 10
10%15=10 < 12  offset = 0
index is 14
14%15 = 14> 12  offset = 14%15%12 = 14%12 = 2
This is set for multiple virtual crystals.

File: src/buildblock/ML_norm.cxx
Comment (gefeichen) [reply src/buildblock/ML_norm.cxx:1102]: As for this one. I also have one problem: 
From tahereh MIC paper, virtual crystal in axial direction is 1. So the num_rings = (4+1)*2 = 10
I think it's for the last virtual crystal in axial and transaxial.

For mCT, it's (13+1)*4 = 56, however, the outermost virtual crystal is ignored. So the num_rings =55.
it's  not necessary  to have (ra == num_rings - 1) || (rb == num_rings - 1).

Comment (KrisThielemans) [reply src/buildblock/ML_norm.cxx:1102]: ok. I agree that for the mCT this condition is dangerous.

I believe that what this condition attempts to do is to detect if it's a virtual crystal. If so, it shouldn't set the fan-data at all. However, the condition is flawed. What about the following.

Add something like this (but adding 2 more args)
```c++
bool is_virtual_crystal(int a, int ra)
{
   const int a_in_block = a % num_transaxial_crystals_in_block;
   if (a_in_block >= num_non_virtual_transaxial_crystals_in_block)
     return true;
   const int ra_in_block = ra % num_axial_crystals_in_block;
   return (ra_in_block >= num_non_virtual_axial_crystals_in_block)
}
```
Then the condition should become
```c++
  if (is_virtual_crystal(a,ra) || is_virtual_crystal_b,rb))
    continue;
```

File: src/buildblock/ML_norm.cxx
Comment (KrisThielemans) [reply src/buildblock/ML_norm.cxx:1069]: Possibly this came about as @tniknejad added a virtual ring at the end (while the mCT doesn't). Might need a bit more thought on how to handle that.

File: src/buildblock/ML_norm.cxx
Comment (KrisThielemans) [reply src/buildblock/ML_norm.cxx:1104]: from my comment above, index=14 is a virtual crystal in your example, so should not be stored. (the bin-value will be zero, and should not be entered in the data without gaps).

Continuing your example, the first non-virtual crystal in the next block would have index=15 (as zero-based). it should be assigned to new_index=12. Using the formula below without offset
```c++
int new_a = a - a/num_transaxial_crystals_per_block*virtual_transaxial_crystals;
```
we'd get, `new_a = 15 - (15/15)*3 = 12`, which is correct. 

Similarly, if `a=16`, `new_a=16-3=13`. Going to the next block `a=30`, `new_a=30-(30/15)*3=3--6=24`.

So I believe no offset needed in all cases.

File: src/buildblock/ML_norm.cxx
Comment (KrisThielemans) [inline src/buildblock/ML_norm.cxx:1105]:
```suggestion
                        int new_a = a - (a/num_transaxial_crystals_per_block)*virtual_transaxial_crystals;
```
I'd feel safer with brackets.

Comment (gefeichen) [reply src/buildblock/ML_norm.cxx:1105]: `bool is_virtual_crystal(int a, int ra)
{
   const int a_in_block = a % num_transaxial_crystals_in_block;
   if (virtual_crystal_transaxial_after_block){
   if (a_in_block >= num_non_virtual_transaxial_crystals_in_block)
     return true;
   }
   else{
   if (a_in_block <= num_of_virtual_transaxial_crystals_in_block)
     return true;
}
const int ra_in_block = ra % num_axial_crystals_in_block;
if (virtual_crystal_axial_after_block){
   return (ra_in_block >= num_non_virtual_axial_crystals_in_block)
else
    return (ra_in_block <= num_of_virtual_axial_crystals_in_block)
}
}`

it is easy to add **virtual_crystal_transaxial_before_or_after_block**. And it can deal with outermost virtual rings in the axial.
Maybe it's not necessary to do such things now.
I think I can do some tests for mCT with and without virtual crystals.

**Comment by KrisThielemans** [comment]
@gefeichen @tniknejad shall we discuss this in a tcon? Please send me and @NikEfth an email with your availability.

File: src/buildblock/ML_norm.cxx
Comment (KrisThielemans) [reply src/buildblock/ML_norm.cxx:1232]: this does look correct

File: src/utilities/apply_normfactors3D.cxx
Comment (KrisThielemans) [reply src/utilities/apply_normfactors3D.cxx:94]: Let's leave this to #776

**Comment by gefeichen** [comment]
refer to #833

**Pull request closed without merging**
